### PR TITLE
[WEB-4025] replace hard-coded TW tables with build-generated alternative

### PR DESCRIPTION
**Pull request by jamiehenson**
## Jira Ticket Link / Motivation

In https://github.com/ably/ably-ui/pull/503, a fix to get the build size down on repos that use `ably-ui` was to manually precompute theming tables for Tailwind classes. This works, but it stinks for if we want to make modifications in future, plus it clogs up the repo. This repo automates what came earlier.

## Summary of changes

The manually added class pairings have been replaced by a function that spits them out at build time into `computed-colors-dark.json` and `computed-colors-light.json`, which are ignored from the repo and bundled in with the package on release. 

There is also a speed advantage here as we're no longer doing a `.find` lookup to dig up the colour, taking us from O(n) to O(1) time.

## How do you manually test this?

Pull down, ensure you've ran a build step (running Storybook will do this), and then check that all the colours work in Storybook (i.e. on the Colors docs page)

## Merge/Deploy Checklist

- [ ] Written automated tests for implemented features/fixed bugs
- [ ] Rebased and squashed commits
- [ ] Commits have clear descriptions of their changes
- [ ] Checked for any performance regressions

## Frontend Checklist

- [ ] No frontend changes in this PR
- [ ] Added before/after screenshots for changes
- [ ] Tested on different platforms/browsers with [Browserstack](https://www.browserstack.com)
- [ ] Compared with the initial design / our [brand guidelines](https://www.figma.com/file/jTQgyIovrhGBOf4Zrvjvbk/Ably-Linear-Design?node-id=118%3A0)
- [ ] Checked the code for accessibility issues ([VoiceOver User Guide](https://support.apple.com/en-gb/guide/voiceover/welcome/mac))?

## Summary by CodeRabbit

- **New Features**
	- Introduced a new script for generating color variants, enhancing the color management system.
	- Added new color arrays and constants for better modularity and accessibility in styling.
	- Expanded Tailwind CSS configuration to process additional JSON files for improved style application.

- **Bug Fixes**
	- Improved color mapping logic by transitioning to a data-driven approach using external JSON files.

- **Chores**
	- Updated package version and added a new development dependency for improved build processes.
	- Modified `.gitignore` to exclude specific files from version control.

### CODE REVIEW

File: src/core/styles/colors/types.ts [hunk 1]
```diff
@@ -145,3 +170,13 @@ export const colorNames = {
...
+
+export const numericalColors = [
+  neutralColors,
+  orangeColors,
+  yellowColors,
+  greenColors,
+  blueColors,
+  violetColors,
+  pinkColors,
+];
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/types.ts:182]: This is intentional.

File: src/core/styles/colors/utils.ts [hunk 2]
```diff
@@ -1,847 +1,9 @@
...
-  { dark: "group-hover:text-pink-300", light: "group-hover:text-pink-700" },
-  { dark: "group-hover:text-pink-400", light: "group-hover:text-pink-600" },
-  { dark: "group-hover:text-pink-500", light: "group-hover:text-pink-500" },
-  { dark: "group-hover:text-pink-600", light: "group-hover:text-pink-400" },
-  { dark: "group-hover:text-pink-700", light: "group-hover:text-pink-300" },
-  { dark: "group-hover:text-pink-800", light: "group-hover:text-pink-200" },
-  { dark: "group-hover:text-pink-900", light: "group-hover:text-pink-100" },
-];
+// If missing, run any build script involving build:colors, i.e. yarn storybook
+import computedColors from "../../../../computed-colors.json";
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/utils.ts:4]: This won't work, utils.ts is a client use-case, not a server use-case

File: src/core/styles/colors/utils.ts [hunk 3]
```diff
@@ -853,5 +15,6 @@
 ) =>
   baseTheme === currentTheme
     ? color
-    : colorMappings.find((set) => set[baseTheme] === color)?.[currentTheme] ||
-      color;
+    : (computedColors as ComputedColors).find(
+        (set) => set[baseTheme] === color,
+      )?.[currentTheme] || color;
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/utils.ts:20]: A neat idea, but unfortunately we can't provide a consistent value for the key unless we double up entries for both forms of theme

[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/utils.ts:20]: I'm not _completely_ against this

Comment (jamiehenson) [reply src/core/styles/colors/utils.ts]: I've done this with two separate files, good suggestion - thank you

File: src/core/styles/colors/utils.ts [hunk 4]
```diff
@@ -1,847 +1,8 @@
...
-];
-
-const colorMappings = [
-  ...textColorMappings,
-  ...bgColorMappings,
-  ...additionalColorMappings,
-];
+// If missing, run any build script involving build:colors, i.e. yarn storybook
+import computedColorsDark from "../../../../computed-colors-dark.json";
+import computedColorsLight from "../../../../computed-colors-light.json";
```
[the remark this answers is not in the document]
Comment (jamiehenson) [reply src/core/styles/colors/utils.ts:5]: Nope, that's enough rabbit, thank you.

File: tailwind.config.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 module.exports = {
-  content: ["./src/**/*.{js,ts,tsx,mdx,stories.tsx}"],
+  content: ["./src/**/*.{js,ts,tsx,mdx,stories.tsx,json}"],
```
Comment (kennethkalmer) [inline tailwind.config.js:2]: Neat, so tailwind will scan this json file looking for class names?

File: scripts/compute-colors.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    ]);
+    console.log(
+      `🎨 Tailwind theme classes have been computed and written to JSON files.`,
+    );
+  } catch {
+    console.error(`Error persisting computed colors.`);
+  }
+}
+
+writeComputedColors();
```
Comment (kennethkalmer) [inline scripts/compute-colors.ts:72]: No `await`? Might just be my lack of async/await knowledge O_o

**Review by kennethkalmer — COMMENTED** [review]
First a comment, otherwise this looks pretty darn good 👏

File: tailwind.config.js (hunk 5 above)
Comment (jamiehenson) [reply tailwind.config.js:2]: Yep, it's all a little bit arcane but it needs to have visibility over these generated JSON files

File: scripts/compute-colors.ts (hunk 6 above)
Comment (jamiehenson) [reply scripts/compute-colors.ts:72]: We don't depend on anything after this so don't need to `await` here - I believe the `async` on this function is to facilitate the `await`s inside it

Comment (kennethkalmer) [reply scripts/compute-colors.ts:72]: Ah, I thought node would just exit before the function finished

**Review by kennethkalmer — APPROVED** [review]
:shipit: (or 🐇it)

**Pull request merged**